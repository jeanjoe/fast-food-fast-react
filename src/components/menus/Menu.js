import React, { Component } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { getMenus, postUserOrder } from "../../actions/menuActions";
import { Container, Item, Icon, Message, Header } from "semantic-ui-react";
import Alert from "../notifications/Alert";
import MenuItem from "./MenuItem";

export class Menu extends Component {
  constructor(props) {
    super(props);
    this.state = {
      loading: true,
      menus: [],
      selectOption: 1,
      location: null,
      ordering: false,
      orderSuccess: null,
      orderError: null
    };
  }

  componentDidMount() {
    const { getMenus } = this.props;
    if (localStorage.getItem("auth_token")) {
      getMenus();
    } else this.props.history.push("/sign-in");
  }

  handleOnChange = event => {
    this.setState({ selectOption: event.target.textContent });
  };

  handleSubmit = id => {
    const { location, selectOption } = this.state;
    const { postUserOrder } = this.props;
    const data = {
      menu_id: id,
      quantity: parseInt(selectOption),
      location: location
    };
    this.setState({ ordering: true });
    postUserOrder(data);
  };

  handleLocation = event => {
    this.setState({ location: event.target.value });
  };

  componentWillReceiveProps(nextProps) {
    this.setState({ loading: false, ordering: false });
    if (nextProps.menus.menus) {
      this.setState({ menus: nextProps.menus.menus });
    } else if (nextProps.error.response.status === 401) {
      this.props.history.push("/sign-in");
    }
    if (nextProps.order.data === "Order Inserted Successfully") {
      this.setState({ orderSuccess: nextProps.order.data });
    }
    if (nextProps.orderError.response) {
      this.setState({ orderError: "Error while placing your order" });
    }
  }

  render() {
    const { loading, menus, ordering, orderSuccess, orderError } = this.state;
    return (
      <Container style={{ padding: "4em 0em" }}>
        <Header as="h2">Menu Items [{menus.length}]</Header>
        {ordering ? <Alert message="Ordering..." type="positive" /> : ""}
        {orderSuccess ? <Message positive>{orderSuccess}</Message> : ""}
        {orderError ? <Message negative>{orderError}</Message> : ""}
        {loading ? (
          <Message icon>
            <Icon name="circle notched" loading />
            <Message.Content>
              <Message.Header>Just one second</Message.Header>
              We are fetching menu items for you.
            </Message.Content>
          </Message>
        ) : (
          <Item.Group>
            {menus.map((menu, key) => (
              <MenuItem
                key={key}
                menu={menu}
                handleLocation={this.handleLocation}
                handleOnChange={this.handleOnChange}
                handleSubmit={this.handleSubmit}
              />
            ))}
          </Item.Group>
        )}
      </Container>
    );
  }
}

Menu.protType = {
  getMenus: PropTypes.func.isRequired,
  postUserOrder: PropTypes.func.isRequired
};

const mapStateToProps = state => ({
  menus: state.menuReducer.menus,
  error: state.menuReducer.errors,
  order: state.menuReducer.order,
  orderError: state.menuReducer.orderError
});

export default connect(
  mapStateToProps,
  { getMenus, postUserOrder }
)(Menu);
