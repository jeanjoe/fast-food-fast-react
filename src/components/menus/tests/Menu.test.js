import React from "react";
import expect from "expect";
import Adapter from "enzyme-adapter-react-16";
import { shallow, configure } from "enzyme";
import { Menu } from "../Menu";
configure({ adapter: new Adapter() });

describe("Test article component", () => {
  const props = {
    history: {
      push: jest.fn()
    },
    getMenus: jest.fn(),
    postUserOrder: jest.fn()
  };
  const wrapper = shallow(<Menu {...props} />);

  it("Matches the snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("Component should mount and redirect if not authenticated", () => {
    wrapper.instance().componentDidMount();
    expect(props.history.push).toHaveBeenCalled();
  });

  it("Component should mount and get menus if authenticated", () => {
    localStorage.setItem("auth_token", "some-token");
    wrapper.instance().componentDidMount();
    expect(props.getMenus).toHaveBeenCalled();
  });

  it("Handle handleOnChange", () => {
    const event = {
      target: {
        textContent: "text"
      }
    };
    wrapper.instance().handleOnChange(event);
    expect(wrapper.state("selectOption")).toEqual("text");
  });

  it("Handle handleLocation", () => {
    const event = {
      target: {
        value: "text"
      }
    };
    wrapper.instance().handleLocation(event);
    expect(wrapper.state("location")).toEqual("text");
  });

  it("Handle handleSubmit", () => {
    const id = 1;
    wrapper.instance().handleSubmit(id);
    expect(props.postUserOrder).toHaveBeenCalled();
  });

  it("Handle componentWillReceiveProps", () => {
    const nextProps = {
      menus: {
        menus: [
          {
            id: 1,
            title: "test title",
            description: "test description",
            price: 1000
          }
        ]
      },
      order: {},
      error: {
        response: {}
      },
      orderError: {}
    };
    wrapper.instance().componentWillReceiveProps(nextProps);
    expect(wrapper.state("menus")).toEqual(nextProps.menus.menus);
  });

  it("Handle componentWillReceiveProps on unauthorizes", () => {
    const nextProps = {
      menus: {
        menus: []
      },
      order: {},
      error: {
        response: {
          status: 401
        }
      },
      orderError: {
        response: {
          status: 401
        }
      }
    };
    wrapper.instance().componentWillReceiveProps(nextProps);
    expect(props.history.push).toHaveBeenCalled();
  });
  it("Handle componentWillReceiveProps when order is inserted", () => {
    const nextProps = {
      menus: {
        menus: []
      },
      order: {
        data: "Order Inserted Successfully"
      },
      error: {
        response: {}
      },
      orderError: {
        response: {}
      }
    };
    wrapper.instance().componentWillReceiveProps(nextProps);
    expect(wrapper.state("orderSuccess")).toEqual(
      "Order Inserted Successfully"
    );
  });
});
