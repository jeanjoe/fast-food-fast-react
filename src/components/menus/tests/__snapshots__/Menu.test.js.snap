// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test article component Matches the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Menu
    getMenus={[MockFunction]}
    history={
      Object {
        "push": [MockFunction] {
          "calls": Array [
            Array [
              "/sign-in",
            ],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": undefined,
            },
          ],
        },
      }
    }
    postUserOrder={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <Header
          as="h2"
        >
          Menu Items [
          0
          ]
        </Header>,
        "",
        "",
        "",
        <Message
          icon={true}
        >
          <Icon
            as="i"
            loading={true}
            name="circle notched"
          />
          <MessageContent>
            <MessageHeader>
              Just one second
            </MessageHeader>
            We are fetching menu items for you.
          </MessageContent>
        </Message>,
      ],
      "style": Object {
        "padding": "4em 0em",
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "as": "h2",
          "children": Array [
            "Menu Items [",
            0,
            "]",
          ],
        },
        "ref": null,
        "rendered": Array [
          "Menu Items [",
          0,
          "]",
        ],
        "type": [Function],
      },
      "",
      "",
      "",
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <Icon
              as="i"
              loading={true}
              name="circle notched"
            />,
            <MessageContent>
              <MessageHeader>
                Just one second
              </MessageHeader>
              We are fetching menu items for you.
            </MessageContent>,
          ],
          "icon": true,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "as": "i",
              "loading": true,
              "name": "circle notched",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": Array [
                <MessageHeader>
                  Just one second
                </MessageHeader>,
                "We are fetching menu items for you.",
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": "Just one second",
                },
                "ref": null,
                "rendered": "Just one second",
                "type": [Function],
              },
              "We are fetching menu items for you.",
            ],
            "type": [Function],
          },
        ],
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Header
            as="h2"
          >
            Menu Items [
            0
            ]
          </Header>,
          "",
          "",
          "",
          <Message
            icon={true}
          >
            <Icon
              as="i"
              loading={true}
              name="circle notched"
            />
            <MessageContent>
              <MessageHeader>
                Just one second
              </MessageHeader>
              We are fetching menu items for you.
            </MessageContent>
          </Message>,
        ],
        "style": Object {
          "padding": "4em 0em",
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "as": "h2",
            "children": Array [
              "Menu Items [",
              0,
              "]",
            ],
          },
          "ref": null,
          "rendered": Array [
            "Menu Items [",
            0,
            "]",
          ],
          "type": [Function],
        },
        "",
        "",
        "",
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <Icon
                as="i"
                loading={true}
                name="circle notched"
              />,
              <MessageContent>
                <MessageHeader>
                  Just one second
                </MessageHeader>
                We are fetching menu items for you.
              </MessageContent>,
            ],
            "icon": true,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "as": "i",
                "loading": true,
                "name": "circle notched",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": Array [
                  <MessageHeader>
                    Just one second
                  </MessageHeader>,
                  "We are fetching menu items for you.",
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "children": "Just one second",
                  },
                  "ref": null,
                  "rendered": "Just one second",
                  "type": [Function],
                },
                "We are fetching menu items for you.",
              ],
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
